### PR TITLE
[ASDataController] Handle moveItem operations

### PR DESCRIPTION
**Pull request by shanereid**
**Note:** This PR is still incomplete, more just for somewhere to put it so @nguyenhuy can review while I iterate on it.

- Also fixes the crash that occurs when you have deletes & moves in the same batchUpdates block.

TODO:
- Re-insert the ASCollectionElements stored in the `movingElements` array using the `movingDestinationIndexPaths` array during the insertion phase.
- Create a test project with a collectionNode with a few known states and make sure each of the states animate properly to each other, each change including deletes, moves and inserts.
- Catch and raise some errors such as a delete and a move source indexpath pointing to the same index, as in a situation where you have a move source and a delete instruction both to the last item in a section you would get a crash anywho.
- Delete commented out code.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/51f30bd0999f9e09001c8851dcf32ed7/raw/acfadc5a3f3579153524392676948243ae18eb17/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/fede33a6825234a67ed42c511146ae6a/raw/909ffe47bad2059146695bdb91b37dff4a402130/log.txt)

File: Source/Details/ASDataController.mm [hunk 1]
```diff
@@ -648,18 +669,36 @@ - (void)_updateElementsInMap:(ASMutableElementMap *)map
...
+                         shouldFetchSizeRanges:shouldFetchSizeRanges
+                                   previousMap:previousMap];
 
-  for (_ASHierarchySectionChange *change in [changeSet sectionChangesOfType:_ASHierarchyChangeTypeDelete]) {
-    NSIndexSet *sectionIndexes = change.indexSet;
-    [map removeSectionsOfItems:sectionIndexes];
+    for (_ASHierarchySectionChange *change in [changeSet sectionChangesOfType:_ASHierarchyChangeTypeDelete]) {
+      NSIndexSet *sectionIndexes = change.indexSet;
+      [map removeSectionsOfItems:sectionIndexes];
+    }
```
Comment (nguyenhuy) [inline Source/Details/ASDataController.mm]: Shouldn't section deletes be processed outside of this `if` block?

File: Source/Details/ASDataController.mm [hunk 2]
```diff
@@ -668,7 +707,20 @@ - (void)_updateElementsInMap:(ASMutableElementMap *)map
                               indexPathsAreNew:YES
                          shouldFetchSizeRanges:shouldFetchSizeRanges
                                    previousMap:previousMap];
+     */
   }
+  
+  [self _insertElementsIntoMap:map kind:ASDataControllerRowNodeKind atIndexPaths:insertChangeIndexPaths
```
Comment (nguyenhuy) [inline Source/Details/ASDataController.mm]: I think we need to sort `insertChangeIndexPaths` in ascending order here.

File: Source/Private/_ASHierarchyChangeSet.mm [hunk 3]
```diff
@@ -99,6 +102,10 @@ + (NSString *)smallDescriptionForItemChanges:(NSArray<_ASHierarchyItemChange *>
 + (void)ensureItemChanges:(NSArray<_ASHierarchyItemChange *> *)changes ofSameType:(_ASHierarchyChangeType)changeType;
 @end
 
+@interface _ASHierarchyItemMoveChange ()
+- (instancetype)initWithSource:(NSIndexPath *)sourceIndexPath andDestination:(NSIndexPath*)destinationIndexPath animationOptions:(ASDataControllerAnimationOptions)animationOptions;
```
Comment (nguyenhuy) [inline Source/Private/_ASHierarchyChangeSet.mm]: To follow the existing convention, let's use:

```objective-c
- (instancetype)initWithSourceIndexPath:(NSIndexPath *)sourceIndexPath andDestinationIndexPath:(NSIndexPath*)destinationIndexPath animationOptions:(ASDataControllerAnimationOptions)animationOptions;
```

File: Source/Private/_ASHierarchyChangeSet.mm [hunk 4]
```diff
@@ -422,11 +434,9 @@ - (void)reloadSections:(NSIndexSet *)sections animationOptions:(ASDataController
...
 - (void)moveItemAtIndexPath:(NSIndexPath *)indexPath toIndexPath:(NSIndexPath *)newIndexPath animationOptions:(ASDataControllerAnimationOptions)options
 {
-  /**
-   * TODO: Proper move implementation.
-   */
-  [self deleteItems:@[ indexPath ] animationOptions:options];
-  [self insertItems:@[ newIndexPath ] animationOptions:options];
+    [self _ensureNotCompleted];
+    _ASHierarchyItemMoveChange *change = [[_ASHierarchyItemMoveChange alloc] initWithSource:indexPath andDestination:newIndexPath animationOptions:options];
+    [_moveItemChanges addObject:change];
```
Comment (nguyenhuy) [inline Source/Private/_ASHierarchyChangeSet.mm]: Nit: Indentation is off.

File: Source/Private/_ASHierarchyChangeSet.mm [hunk 5]
```diff
@@ -578,6 +588,40 @@ - (void)_validateUpdate
...
+      ASFailUpdateValidation(@"Attempt to move item %zd from section %zd into section %zd, but there are only %zd sections after the update.", sourceItem, sourceSection, destinationSection, newSectionCount);
+    }
+    
+    // Assert that the move destination is to a valid post-insert item
+    NSInteger newItemCount = _newItemCounts[destinationSection];
+    if (destinationItem >= newItemCount) {
+      ASFailUpdateValidation(@"Attempt to insert item %zd from section %zd into position %zd within section %zd, which only contains %zd items after the update.", sourceItem, sourceSection, destinationItem, destinationSection, newItemCount);
+      return;
+    }
+  }
```
Comment (nguyenhuy) [inline Source/Private/_ASHierarchyChangeSet.mm]: Ha, this validation step is cleaner than I anticipated! 👍

File: Source/Private/_ASHierarchyChangeSet.mm [hunk 6]
```diff
@@ -1015,3 +1062,21 @@ - (NSString *)debugDescription
 }
 
 @end
+
+@implementation _ASHierarchyItemMoveChange
+- (instancetype)initWithSource:(NSIndexPath *)sourceIndexPath andDestination:(NSIndexPath *)destinationIndexPath animationOptions:(ASDataControllerAnimationOptions)animationOptions
```
Comment (nguyenhuy) [inline Source/Private/_ASHierarchyChangeSet.mm]: Nit: Add a new line above this line.

**Comment by harryworld** [comment]
Regarding to the 3rd TODO, I wonder if this is appropriate.

Based on the UIKit implementation, if you delete and move from the same index path, you will get the following error

```
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to perform a delete and a move from the same index path (<NSIndexPath: 0x6180000291c0> {length = 2, path = 0 - 1})'
```

The current Texture does not call the UIKit `moveItemAtIndexPath:toIndexPath:` and thus not throwing the exception. However, it would also mean the native move animation is not provided.

Please see the [sample project](https://github.com/harryworld/CollectionViewDeleteMoveFrom) for the crash.
A [stackoverflow link](https://stackoverflow.com/a/38625897/2332394) here makes a good point about UIKit

`
A handy way to verify the operations are set up correctly: the set of reload, insert and move-to indexPaths should not have any duplicates, and the set of reload, delete, and move-from indexPaths should not have any duplicates.
`

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Any source code changes should have an entry in CHANGELOG.md or have #trivial in their title.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/0d7e5eb33382180a800cb43b2d631cbd/raw/5df34903f7df2a0d7ba8f74224c0ea00f360b9b5/log.txt)

**Comment by nguyenhuy** [comment]
Given that there hasn't been any progress on this diff, I'm going to close it. Please open a new one if need to. Thank you.

**Pull request closed without merging**